src/net_doh.c

```diff
@@ -0,0 +1,1070 @@
+    }
+    if (n < 0) {
+        int err = SSL_get_error(self->ssl, n);
+        switch (err) {
+        case SSL_ERROR_WANT_READ:
+            errno = EAGAIN;
+            break;
+        case SSL_ERROR_SYSCALL:
+            switch (errno) {
+            case ECONNREFUSED:
+            case ECONNRESET:
+            case ENOTCONN:
+                perf__doh_reconnect(sock);
+                errno = EAGAIN;
+                break;
+            default:
+                break;
+            }
+            break;
+        default:
+            errno = EBADF;
+            break;
+        }
+        PERF_UNLOCK(&self->lock);
+        return -1;
+    }
+
+    PERF_UNLOCK(&self->lock);
+
+    // make sure we can process data
+    if (self->is_ready &&
+        self->http2 != NULL) {
+        // this will be processed by nghttp2 callbacks
+        // self->recvbuf holds the payload
+        PERF_LOCK(&self->lock);
+
+        ret = nghttp2_session_mem_recv(self->http2->session, 
+                                       (uint8_t*) self->recvbuf, 
+                                       n);
+     
+        if (ret < 0) {
+            // 
+            perf_log_warning("nghttp2_session_mem_recv failed: %s", 
+                            nghttp2_strerror((int) ret));
+
+            http2_session_free(sock);
+            PERF_UNLOCK(&self->lock);     
+            return -1;
+        }
+
+        // need to execute nghttp2_session_send if the receive ops triggered data frames 
+        ret = nghttp2_session_send(self->http2->session);
+        if (ret < 0) {
+            perf_log_warning("nghttp2_session_send failed: %s", nghttp2_strerror((int) ret));
+            PERF_UNLOCK(&self->lock);
+            return -1;
+        }
+
+        PERF_UNLOCK(&self->lock);
+    }
+    
+    PERF_LOCK(&self->lock);
+
+    if (self->http2 &&
+        self->http2->dnsmsg_completed) {
+        if (self->http2->dnsmsg_at > len) {
+            perf_log_warning("failed to process result - DNS response size");
+            PERF_UNLOCK(&self->lock);
+            return -1;
+        }
+
+        memcpy(buf, self->http2->dnsmsg, self->http2->dnsmsg_at);
+
+        self->http2->dnsmsg_completed = false;
+        int response_len = self->http2->dnsmsg_at;
+        self->http2->dnsmsg_at = 0;
+
+        self->have_more = false;
+        PERF_UNLOCK(&self->lock);
+        return response_len;
+    } else {
+        self->have_more = true;
+        PERF_UNLOCK(&self->lock);
+        errno = EAGAIN;
+        return -1;
+    }
+}
+
+static ssize_t perf__doh_sendto(struct perf_net_socket* sock, uint16_t qid, const void* buf, size_t len, int flags, const struct sockaddr* dest_addr, socklen_t addrlen)
+{
+    int ret = -1;
+
+    PERF_LOCK(&self->lock);
+
+    if (!self->is_ready) {
+        PERF_UNLOCK(&self->lock);
+        errno = EINPROGRESS;
+        return -1;
+    }
+
+    self->qid = qid;
+ 
+    if (self->is_ready) {
+        if (memcmp(net_doh_method, "GET", 3) == 0) {
```

Comment:
`net_doh_method` should be converted to a `enum` in `sock` during `opensocket`, then these `memcmp` can be replaced with a `switch` and the fatal check below can be performed at start and not every send.